src/read.c

```diff
@@ -1353,86 +1339,93 @@ static avifCodecType avifDecoderItemGetGridCodecType(const avifDecoderItem * gri
-            if (tileCodecType == AVIF_CODEC_TYPE_UNKNOWN) {
-                continue;
-            }
-            if (item->hasUnsupportedEssentialProperty) {
-                // An essential property isn't supported by libavif; can't
-                // decode a grid image if any tile in the grid isn't supported.
-                avifDiagnosticsPrintf(&decoder->diag, "Grid image contains tile with an unsupported property marked as essential");
+        if (item->dimgForID != gridItem->id) {
+            continue;
+        }
+
+        // According to HEIF (ISO 14496-12), Section 6.6.2.3.1, the SingleItemTypeReferenceBox of type 'dimg'
+        // identifies the input images of the derived image item of type 'grid'. Since the reference_count
+        // shall be equal to rows*columns, unknown tile item types cannot be skipped but must be considered
+        // as errors.
+        const avifCodecType tileCodecType = avifGetCodecType(item->type);
+        if (tileCodecType == AVIF_CODEC_TYPE_UNKNOWN) {
+            avifDiagnosticsPrintf(&decoder->diag, "Tile item ID %u has an unknown item type '%.4s'", item->id, (const char *)item->type);
+            return AVIF_FALSE;
+        }
+
+        if (item->hasUnsupportedEssentialProperty) {
+            // An essential property isn't supported by libavif; can't
+            // decode a grid image if any tile in the grid isn't supported.
+            avifDiagnosticsPrintf(&decoder->diag, "Grid image contains tile with an unsupported property marked as essential");
+            return AVIF_FALSE;
+        }
+
+        const avifTile * tile =
+            avifDecoderDataCreateTile(decoder->data, tileCodecType, item->width, item->height, avifDecoderItemOperatingPoint(item));
+        if (!tile) {
+            return AVIF_FALSE;
+        }
+        if (!avifCodecDecodeInputFillFromDecoderItem(tile->input,
+                                                     item,
+                                                     decoder->allowProgressive,
+                                                     decoder->imageCountLimit,
+                                                     decoder->io->sizeHint,
+                                                     &decoder->diag)) {
+            return AVIF_FALSE;
+        }
+        tile->input->alpha = alpha;
+
+        if (firstTileItem == NULL) {
+            firstTileItem = item;
+
+            // Adopt the configuration property of the first image item tile, so that it can be queried from
+            // the top-level color/alpha item during avifDecoderReset().
+            const char * configPropName = memcmp(item->type, "av02", 4) ? "av1C" : "av2C";
+            const avifProperty * srcProp = avifPropertyArrayFind(&item->properties, configPropName);
+            if (!srcProp) {
+                avifDiagnosticsPrintf(&decoder->diag, "Grid image's first tile is missing an %s property", configPropName);
```

Comment:
`tile` or its contents need to be cleaned up here and in other locations, or is the cleanup expected when the backing array is destroyed?

Reply:
I see that `avifDecoderDataDestroy()` calls `avifDecoderDataClearTiles()` so it should be fine.